workflow/src/main/java/bio/terra/tanagra/workflow/FlattenHierarchy.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+/**
+ * A batch Apache Beam pipeline for flattening hierarchical parent-child relationships to
+ * ancestor-descendant relationships.
+ */
+// TODO consider how to make this easier to reuse by deployers.
+public final class FlattenHierarchy {
+  private FlattenHierarchy() {}
+
+  /** Options supported by {@link FlattenHierarchy}. */
+  public interface FlattenHierarchyOptions extends BigQueryOptions {
+    @Description(
+        "Path to a BigQuery standard SQL query file to execute to retrieve the hierarchy to be"
+            + "flattened. The result of the query should have two columns, (parent, child) that"
+            + "defines all of the direct relationships of the hierarchy to be flattened.")
+    String getHierarchyQuery();
+
+    void setHierarchyQuery(String query);
+
+    @Description(
+        "The maximum depth of ancestors present than the hierarchy. This may be larger "
+            + "than the actual max depth, but if it is smaller the resulting table will be incomplete")
+    @Default.Integer(64)
+    int getMaxHierarchyDepth();
+
+    void setMaxHierarchyDepth(int val);
+
+    @Description(
+        "The  \"[project_id]:[dataset_id].[table_id]\" specification of the ancestor table to create.")
+    String getOutputBigQueryTable();
+
+    void setOutputBigQueryTable(String tableRef);
+
+    @Description("What to name the ancestor column on the output BigQuery table.")
+    @Default.String("ancestor")
+    String getOutputAncestorColumn();
+
+    void setOutputAncestorColumn(String ancestorColumn);
+
+    @Description("What to name the descendant column on the output BigQuery table.")
+    @Default.String("descendant")
+    String getOutputDescendantColumn();
+
+    void setOutputDescendantColumn(String descendantColumn);
+  }
+
+  /** Parses a {@link TableRow} as a row contianing a (parent_id, child_id) long pair. */
+  // TODO support string ids in addition to longs.
+  private static KV<Long, Long> relationshipRowToKV(TableRow row) {
+    Long parentId = Long.parseLong((String) row.get("parent"));
+    Long childId = Long.parseLong((String) row.get("child"));
+    return KV.of(parentId, childId);
+  }
+
+  /** Defines the BigQuery schema for the output hierarchy table. */
+  private static TableSchema flattenedHierarchySchema(FlattenHierarchyOptions options) {
+    return new TableSchema()
+        .setFields(
+            List.of(
+                new TableFieldSchema()
+                    .setName(options.getOutputAncestorColumn())
+                    .setType("INTEGER"),
+                new TableFieldSchema()
+                    .setName(options.getOutputDescendantColumn())
+                    .setMode("REPEATED")
```

Comment:
TO be clear about what this is doing, we're going to expand the entire subtree for each concept, and put it into this singular column - correct? Or is this only the leaves?

What are the limits on this in BigQuery (or if not there, what is a reasonable limit in general?). I believe that many of these trees are quite large (20K concepts, maybe more)

Reply:
Yep, that's correct. `ancestor, array[descendants]`
The BigQuery limit is 100MB per row. The synpuf dataset has 100k children for the largest  condition snomed concept, 'Clinical Finding'.
Maybe not such a great idea to group it this way. The idea with this table is to support the 'DESCENDANT_OF' operator. With this structuring, after the scan on the ancestor column, we only need to look at a single descendant column.
 Some alternatives, would love to hear other suggestions:
single `ancestor, descendant` per row, so many duplicates of a given ancestor.
descendant to array[ancestor]. at lookup time, we need to scan the entire descendants table to look for ancestors. 
thoughts?

Reply:
Also @freemabd , what is the largest concept tree the cohort builder currently supports? IIRC you actually disable some of the larger nodes, and do not let the user select them (I think "clinical finding" probably falls in this category)

Reply:
`ancestor, descendant` SGTM.
Clustering will have to wait a bit since it's disabled without a time partition due to an out of date biquery restriction :/ But there's a PR just out, so there's hope.
https://issues.apache.org/jira/browse/BEAM-10652

Reply:
> Also @freemabd , what is the largest concept tree the cohort builder currently supports? IIRC you actually disable some of the larger nodes, and do not let the user select them (I think "clinical finding" probably falls in this category)

Definitely, SNOMED Conditions tree. Clinical finding is the very top level of that tree, but everything under that node is selectable. So the largest nodes are probably "Disease" and "Finding by site" with ~28K descendants